index.d.ts

```diff
@@ -1,49 +1,68 @@
-    {
-        jwt: Jwt;
-    }
+import * as fastify from 'fastify';
+import * as http from 'http';
+import * as jwt from 'jsonwebtoken';
+
+declare module 'fastify' {
+  namespace JWTTypes {
+    type PayloadType = object | string;
+
+    interface SignCallback extends jwt.SignCallback {}
 
-    interface FastifyRequest<
-        HttpRequest,
-        Query = DefaultQuery,
-        Params = DefaultParams,
-        Headers = DefaultHeaders,
-        Body = DefaultBody
-        >
-    {
-        jwtVerify(callback?: VerifyCallback): Promise<any>;
+    interface VerifyCallback<Decoded extends PayloadType> extends jwt.VerifyCallback {
+      (err: jwt.VerifyErrors, decoded: Decoded): void;
     }
+  }
+
+  interface JWT {
+    options: {
+      decode: jwt.DecodeOptions;
+      sign: jwt.SignOptions;
+      verify: jwt.VerifyOptions;
+    };
+    secret: jwt.Secret;
+
+    sign(payload: JWTTypes.PayloadType | Buffer, options?: jwt.SignOptions): string;
+    sign(payload: JWTTypes.PayloadType | Buffer, callback: JWTTypes.SignCallback): void;
+    sign(payload: JWTTypes.PayloadType | Buffer, options: jwt.SignOptions, callback: JWTTypes.SignCallback): void;
+
+    verify<Decoded extends JWTTypes.PayloadType = any>(token: string, options?: jwt.VerifyOptions): Decoded;
+    verify<Decoded extends JWTTypes.PayloadType = any>(token: string, callback: JWTTypes.VerifyCallback<Decoded>): void;
+    verify<Decoded extends JWTTypes.PayloadType = any>(
+      token: string,
+      options: jwt.VerifyOptions,
+      callback: JWTTypes.VerifyCallback<Decoded>,
+    ): void;
+
+    decode<Decoded extends JWTTypes.PayloadType = any>(token: string, options?: jwt.DecodeOptions): null | Decoded;
+  }
+
+  interface FastifyInstance {
+    jwt: JWT;
+  }
+
+  interface FastifyReply<HttpResponse> {
+    jwtSign(payload: JWTTypes.PayloadType | Buffer, options?: jwt.SignOptions): Promise<string>;
+    jwtSign(payload: JWTTypes.PayloadType | Buffer, callback: JWTTypes.SignCallback): void;
+    jwtSign(payload: JWTTypes.PayloadType | Buffer, options: jwt.SignOptions, callback: JWTTypes.SignCallback): void;
+  }
+
+  interface FastifyRequest<HttpRequest> {
+    jwtVerify<Decoded extends JWTTypes.PayloadType = any>(options?: jwt.VerifyOptions): Promise<Decoded>;
+    jwtVerify<Decoded extends JWTTypes.PayloadType = any>(callback: JWTTypes.VerifyCallback<Decoded>): void;
+    jwtVerify<Decoded extends JWTTypes.PayloadType = any>(
+      options: jwt.VerifyOptions,
+      callback: JWTTypes.VerifyCallback<Decoded>,
+    ): void;
+  }
+}
+
+declare interface FastifyJWTOptions {
+  secret: jwt.Secret;
+  decode?: jwt.DecodeOptions;
+  sign?: jwt.SignOptions;
+  verify?: jwt.VerifyOptions;
 }
 
-declare const fastifyJWT: fastify.Plugin<Server, IncomingMessage, ServerResponse, { secret: Secret }>;
+declare const fastifyJWT: fastify.Plugin<any, any, any, FastifyJWTOptions>;

```

Comment:
Concerned with the use of `any` here. Why was this changed from the `http` items? I'm unsure how the `fastify.Plugin` type signature works so maybe something needs to be changed there too.

Reply:
`HttpServer` / `HttpRequest` / `HttpResponse` in `fastify.Plugin<HttpServer, HttpRequest, HttpResponse, T>` maybe `http2` type (See [here](https://github.com/fastify/fastify/blob/b23a2a3923e6b644e910c89d4cabd1536c81922d/fastify.d.ts#L12-L14)).

However, this plugin don't require there type information, so I set `any` type.

Reply:
Leave this as is for now. I'm working on making the base `fastify` types better so we will introduce a better fix here later on

Reply:
Ok, revert it.